### ISSUE TITLE
building from source on linux (trying to create an official debian package)

### ISSUE THREAD

**Issue by alexmyczko** [be5invis/Iosevka#715]
The problem is npm during build downloads stuff. How easy would it be to turn the downloads off and still use the same build system? How hard to switch to another build system?

http://phd-sid.ethz.ch/debian/fonts-iosevka/

(Debian builds may not access internet during build)

**Comment by alerque** [comment]
Fontship would be one alternative build system for this, but that isn't packaged on Debian yet either. The build tooling for this repo is considerably more complicated than your average Font repo. Switching it out for something else will not be trivial.

NPM allows for offline usage by specifying a module cache dir and warming it up with the necessary packages in a pre-build step, then builds can happen offline. I know Debian does this for other packages (as does Arch Linux).

**Comment by alexmyczko** [comment]
Ok fontship you say here we go `fontship -V` says `fontship v0.6.2`, it was fun to build from source...
with npm build i'd download one half of the internet, and to build fontship (cargo) it would download the other half of the internet.

ok, now what?

**Comment by alerque** [comment]
Ha ha. You forgot about the half that `pip` would need to download ;-)

Now what? This font project isn't arranged quite right to use it yet. Someday I'll look at porting the current build rules for the various flavors of the font to Fontship rules. I wasn't proposing you could build this font with it now, just that it is an alternative build system that could be considered for the project. Just like many C++ projects could be built with either CMake or GNU Make or Meson but projects typically only support one of the above, so for font projects. Most projects use totally hand baked (or copy pasted) build scripts and a hodge-podge of Python stuff. With Fontship I'm trying to organize that a bit into best practices for automated font builds so that a bunch of Mac based designer types don't have to keep re-inventing the wheel on the code side.

This font project is a bit more advanced than most though, so it's not the one I'm starting with.

**Comment by be5invis** [comment]
It may be possible to do a rollup for all the JS it uses into a single file, and running it via with Node.JS will create a font (and then you could run other stages, like creating WOFF2 or TTFAutohint).
Put into 4.x.

**Comment by alexmyczko** [comment]
I would only need to build ttf with autohint, or better otf but i could not find otf building support..

**Comment by be5invis** [comment]
@alexmyczko I do not think I have enough bandwidth on that.

**Comment by alexmyczko** [comment]
meanwhile i missed the bookworm freeze, while others have it https://repology.org/project/fonts:iosevka/versions

**Comment by hackerb9** [comment]
> meanwhile i missed the bookworm freeze, while others have it https://repology.org/project/fonts:iosevka/versions
> 
> and my sid.ethz.ch repo has 22.0.0 now

The link from the Iosevka README.md goes to phd-sid.ethz.ch which is stuck on version 7. Is that fixable? Thanks!

**Comment by alexmyczko** [comment]
@hackerb9 the link really is sid.ethz.ch, please don't call it phd-sid (this is an internal name that can change any time, i'm not authority for it, but sid is mine)

Anyways, the main problem is that our web server is not configured to sort the way convenient for human beings, but if you do this:

`links2 -dump sid.ethz.ch/debian/fonts-iosevka |sort -V`

it's a bit better and shows there's packaging for 22.0.0. for transparency and efficiency (not wasting time removing old stuff), i keep old stuff around (which i look up for reasons)

HTH

**Comment by hackerb9** [comment]
Okay. I think I understand now. 

The [README for Iosevka](https://github.com/be5invis/Iosevka?tab=readme-ov-file#installing-via-package-managers) directly links to _phd-sid_. The code reads:
```
 * Ubuntu Linux: Install one of the [fonts-iosevka packages](http://phd-sid.ethz.ch/debian/fonts-iosevka/).
```

Is it worth filing a new issue for this with the Iosevka project?

(By the way, thank you for creating a Debian package! I hope it gets accepted into the official repository soon.)

**Comment by alexmyczko** [comment]
https://github.com/be5invis/Iosevka/issues/715#issuecomment-723037191

it will not unless someone fixes ghe building part. building is building, not downloading stuff from internet (forbidden).

**Comment by hackerb9** [comment]
> [#715 (comment)](https://github.com/be5invis/Iosevka/issues/715#issuecomment-723037191)
> 
> it will not unless someone fixes the building part. building is building, not downloading stuff from internet (forbidden).

While I certainly would prefer a package that generates the font, I thought that wasn't a strict requirement for  Debian. For example, `fonts-terminus` (which is under the SIL Open Font License) just copies the .ttf.

The [Debian Font Packaging Policy](https://wiki.debian.org/Fonts/PackagingPolicy) Wiki does not directly mention building but does link to a draft of a [Debian fonts-policy](https://github.com/rbrito/fonts-policy/blob/master/fonts_policy.mdwn) which reads:

> Generation of binary packages
>
>To preserve the forms of the fonts as intended by the original authors, packagers MAY opt, for fonts distributed in a final, usable form (e.g., .ttf, .otf etc.), to ship those files in the binary packages, without any build step.
>
>For packages that only provide sources in [Fontforge](http://fontforge.sf.net/)'s format, custom scripts MAY be used, if upstream does not provide them.
>
>Despite the suggestion of shipping fonts that are as close as possible to what upstream ships, some manipulation is allowed and even encouraged, for the removal of glyphs for reduction of the size to fit in installation media.
>
>In any case, the package SHOULD clearly state that the font was manipulated in the README.Debian file of the generated binary package.

Would it be possible to release an interim package using the binary files, at least until the building part works?

**Comment by alexmyczko** [comment]
Yes that is absolutely possible, but then I prefer the fnt tool which works on Debian and macOS, and according the issues there is a plan to have own sources as in any URL... https://github.com/alexmyczko/fnt

I will need some time, what is frustrating is building it from source is good, but I don't remember what was not good about it. It can not easily be fixed?

**Comment by be5invis** [comment]
If you look into the `DistUnhintedTTF` recipe in `verdafile` you could find it eventually does a "function call" to `packages/font/src/index.mjs` with some parameters as a JSON object. Maybe you can try capture it and do it with a rollup script that bundled version of `packages/font/src/index.mjs`...

**Comment by josch** [comment]
@alexmyczko the link seems to be 404 right now. Can this be fixed? Thanks!

**Comment by alexmyczko** [comment]
@josch sorry about the outage, it's back

**Comment by alexmyczko** [comment]
just for the record, it has been found too big to package the font directly. and source build would be preferred, but as long as "verda" is not packaged, this won't happen (not from me)

**Comment by alexmyczko** [comment]
new URL for deb src pkg: https://github.com/be5invis/Iosevka/issues/2873

**Comment by alexmyczko** [comment]
meanwhile, 5 years later nobody cared to package node-verda...

@be5invis ok so `npm install verda` can be put into the package, offline:
it's only about 319 MB, and 14558 files that require a machine readable copyright file, like https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/ that nobody wants to write manually...

i'll try to generate something with awk or so... it'll need to pass this place:
https://salsa.debian.org/ftp-team/manpages/-/blob/master/doc/ftp-new.7.man?ref_type=heads

```
node_modules$ licensecheck --recursive . | sed s/.\*:\ //g | sort | uniq -c
     16 *No copyright* Apache License 2.0
      1 *No copyright* BSD 0-Clause License
     21 *No copyright* BSD 2-Clause License
      8 *No copyright* BSD 3-Clause License
      1 *No copyright* Boost Software License
      1 *No copyright* GNU General Public License
     31 *No copyright* ISC License
    270 *No copyright* MIT License
      4 *No copyright* MIT License and/or X11 License
      1 *No copyright* Python License 2.0
   9554 *No copyright* UNKNOWN
      4 *No copyright* UNKNOWN [generated file]
      1 Apache License 2.0
      2 Apache License 2.0 and/or BSD 2-Clause License
      7 BSD 0-Clause License
    105 BSD 2-Clause License
      4 BSD 2-Clause with views sentence
     47 BSD 3-Clause License
      1 BSD 3-Clause License and/or MIT License
      1 Boost Software License
      1 Do What The Fuck You Want To Public License, Version 2
     33 ISC License
      1 ISC License and/or MIT License
    233 MIT License
      4 MIT License [generated file]
      1 MIT License and/or X11 License
    914 UNKNOWN
      1 UNKNOWN [generated file]
   1254 Unicode License Agreement - Data Files and Software (2016)
```

**Comment by alexmyczko** [comment]
https://tracker.debian.org/news/1719470/accepted-fonts-iosevka-2200ds-1-source-all-into-unstable/

i am closing this
